components/style/matching.rs

```diff
@@ -526,20 +534,16 @@ pub trait ElementMatchMethods<'le> : TElement<'le> {
                                                  style_attribute,
                                                  None,
                                                  &mut applicable_declarations.normal);
-        stylist.push_applicable_declarations(self,
-                                             parent_bf,
-                                             None,
-                                             Some(PseudoElement::Before),
-                                             &mut applicable_declarations.before);
-        stylist.push_applicable_declarations(self,
-                                             parent_bf,
-                                             None,
-                                             Some(PseudoElement::After),
-                                             &mut applicable_declarations.after);
+        Self::Impl::each_eagerly_cascaded_pseudo_element(|pseudo| {
+            stylist.push_applicable_declarations(self,
+                                                 parent_bf,
+                                                 None,
+                                                 Some(pseudo.clone()),
+                                                 applicable_declarations.per_pseudo.entry(pseudo).or_insert(vec![]));
+        });
 
         applicable_declarations.normal_shareable &&
-        applicable_declarations.before.is_empty() &&
-        applicable_declarations.after.is_empty()
+        applicable_declarations.per_pseudo.values().all(|v| v.is_empty())
```

Comment:
We have to be careful here: pseudo-elements that have UA styles with a `*|*` selector risk defeating style sharing completely.

What pseudo-elements are likely to be in `applicable_declarations.per_pseudo` at this point? Only the eagerly-cascaded ones?